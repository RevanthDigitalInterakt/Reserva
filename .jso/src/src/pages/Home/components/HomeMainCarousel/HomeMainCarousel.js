  Object.defineProperty(exports, "__esModule", {
    value: true
  });
  exports.default = undefined;
  var _slicedToArray2 = _$$_REQUIRE(_dependencyMap[0])(_$$_REQUIRE(_dependencyMap[1]));
  var _react = _interopRequireWildcard(_$$_REQUIRE(_dependencyMap[2]));
  var _reactNative = _$$_REQUIRE(_dependencyMap[3]);
  var _reactNativeReanimatedCarousel = _$$_REQUIRE(_dependencyMap[0])(_$$_REQUIRE(_dependencyMap[4]));
  var _CarouselPaginationItem = _$$_REQUIRE(_dependencyMap[0])(_$$_REQUIRE(_dependencyMap[5]));
  var _ImageComponent = _$$_REQUIRE(_dependencyMap[0])(_$$_REQUIRE(_dependencyMap[6]));
  var _CarouselScrollIndicator = _$$_REQUIRE(_dependencyMap[0])(_$$_REQUIRE(_dependencyMap[7]));
  var _configDeviceSizes = _$$_REQUIRE(_dependencyMap[0])(_$$_REQUIRE(_dependencyMap[8]));
  var _testProps = _$$_REQUIRE(_dependencyMap[0])(_$$_REQUIRE(_dependencyMap[9]));
  function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
  function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
  function HomeMainCarousel(_ref) {
    var data = _ref.data;
    var navigation = (0, _$$_REQUIRE(_dependencyMap[10]).useNavigation)();
    var progressValue = (0, _$$_REQUIRE(_dependencyMap[11]).useSharedValue)(0);
    var _useRemoteConfig = (0, _$$_REQUIRE(_dependencyMap[12]).useRemoteConfig)(),
      getBoolean = _useRemoteConfig.getBoolean;
    var _useState = (0, _react.useState)(0),
      _useState2 = (0, _slicedToArray2.default)(_useState, 2),
      currIndex = _useState2[0],
      setCurrIndex = _useState2[1];
    var $carousel = (0, _react.useRef)();
    var slideDelay = (0, _react.useMemo)(function () {
      return (data.showtime || 10) * 1000;
    }, [data.showtime]);
    var showNewHome = getBoolean('show_new_home');
    var carouselHeight = (0, _react.useMemo)(function () {
      var _data$items = (0, _slicedToArray2.default)(data.items, 1),
        item = _data$items[0];
      var _image = item.image,
        width = _image.width,
        height = _image.height;
      return width && height ? height * _configDeviceSizes.default.DEVICE_WIDTH / width : 500;
    }, [data.items]);
    var _onPress = (0, _react.useCallback)(function (item) {
      var _item$filters, _item$filters$priceFi, _item$filters2, _item$filters2$priceF, _item$filters3, _item$filters3$priceF;
      var reservaMini = item.reservaMini,
        orderBy = item.orderBy;
      var navigateParams = {
        facetInput: item.facets,
        referenceId: item.reference,
        reservaMini: reservaMini,
        orderBy: orderBy
      };
      if (((_item$filters = item.filters) != null && (_item$filters$priceFi = _item$filters.priceFilter) != null && _item$filters$priceFi.from || ((_item$filters2 = item.filters) == null ? undefined : (_item$filters2$priceF = _item$filters2.priceFilter) == null ? undefined : _item$filters2$priceF.from) === null) && (_item$filters3 = item.filters) != null && (_item$filters3$priceF = _item$filters3.priceFilter) != null && _item$filters3$priceF.to) {
        var _item$filters4, _item$filters4$priceF, _item$filters5, _item$filters5$priceF;
        navigateParams.filters = {
          priceFilter: {
            from: ((_item$filters4 = item.filters) == null ? undefined : (_item$filters4$priceF = _item$filters4.priceFilter) == null ? undefined : _item$filters4$priceF.from) || 0,
            to: ((_item$filters5 = item.filters) == null ? undefined : (_item$filters5$priceF = _item$filters5.priceFilter) == null ? undefined : _item$filters5$priceF.to) || 0
          }
        };
      }
      return navigation.navigate('ProductCatalog', navigateParams);
    }, [navigation]);
    return /*#__PURE__*/(0, _$$_REQUIRE(_dependencyMap[13]).jsxs)(_reactNative.View, Object.assign({
      style: {
        flex: 1
      }
    }, (0, _testProps.default)('default_carrousel_container'), {
      children: [/*#__PURE__*/(0, _$$_REQUIRE(_dependencyMap[13]).jsx)(_reactNativeReanimatedCarousel.default, Object.assign({}, (0, _testProps.default)('default_carrousel_content'), {
        loop: true,
        width: _configDeviceSizes.default.DEVICE_WIDTH,
        height: carouselHeight,
        ref: function ref(carousel) {
          if (carousel) $carousel.current = carousel;
        },
        mode: "parallax",
        modeConfig: {
          parallaxScrollingScale: 1,
          parallaxScrollingOffset: 50
        },
        enabled: data.items.length > 1,
        onProgressChange: function onProgressChange(_, absoluteProgress) {
          progressValue.value = absoluteProgress;
        },
        panGestureHandlerProps: {
          activeOffsetX: [-10, 10]
        },
        data: data.items,
        style: {
          backgroundColor: 'rgba(0, 0, 0, 0)',
          position: 'relative'
        },
        onSnapToItem: setCurrIndex,
        renderItem: function renderItem(_ref2) {
          var item = _ref2.item;
          return /*#__PURE__*/(0, _$$_REQUIRE(_dependencyMap[13]).jsx)(_$$_REQUIRE(_dependencyMap[14]).Box, {
            alignItems: "flex-start",
            bg: "white",
            children: /*#__PURE__*/(0, _$$_REQUIRE(_dependencyMap[13]).jsx)(_$$_REQUIRE(_dependencyMap[14]).Box, {
              mb: "quarck",
              width: 1,
              children: /*#__PURE__*/(0, _$$_REQUIRE(_dependencyMap[13]).jsx)(_reactNative.Pressable, Object.assign({}, (0, _testProps.default)('carrousel_button'), {
                onPress: function onPress() {
                  return _onPress(item);
                },
                delayLongPress: 100,
                children: /*#__PURE__*/(0, _$$_REQUIRE(_dependencyMap[13]).jsx)(_ImageComponent.default, {
                  style: [{
                    height: carouselHeight
                  }, showNewHome ? {
                    borderBottomLeftRadius: 8,
                    borderBottomRightRadius: 8
                  } : null],
                  resizeMode: "cover",
                  source: {
                    uri: item.image.url
                  }
                })
              }))
            })
          });
        }
      })), showNewHome ? /*#__PURE__*/(0, _$$_REQUIRE(_dependencyMap[13]).jsx)(_reactNative.View, {
        style: [_$$_REQUIRE(_dependencyMap[15]).styles.bulletsWrapper, {
          width: data.items.length * 24.5
        }],
        children: data.items.map(function (item, i) {
          return /*#__PURE__*/(0, _$$_REQUIRE(_dependencyMap[13]).jsx)(_CarouselPaginationItem.default, {
            backgroundColor: _$$_REQUIRE(_dependencyMap[16]).COLORS.WHITE,
            animValue: progressValue,
            index: i,
            actualPosition: currIndex,
            length: data.items.length,
            slideDelay: slideDelay,
            onFinishAnimation: function onFinishAnimation() {
              var _$carousel$current;
              return (_$carousel$current = $carousel.current) == null ? undefined : _$carousel$current.next();
            }
          }, `home-main-carousel-${item.image.url}`);
        })
      }) : /*#__PURE__*/(0, _$$_REQUIRE(_dependencyMap[13]).jsx)(_CarouselScrollIndicator.default, {
        carouselLength: data.items.length,
        actualPosition: currIndex,
        slideDelay: slideDelay,
        onFinishAnimation: function onFinishAnimation() {
          var _$carousel$current2;
          return (_$carousel$current2 = $carousel.current) == null ? undefined : _$carousel$current2.next();
        }
      })]
    }));
  }
  var _default = exports.default = HomeMainCarousel;
