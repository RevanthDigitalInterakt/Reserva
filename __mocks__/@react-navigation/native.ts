const navigate = jest.fn();

const useNavigation = () => ({
  navigate,
});

export {
  useNavigation,
};
