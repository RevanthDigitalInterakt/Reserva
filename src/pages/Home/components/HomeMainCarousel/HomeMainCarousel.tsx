import React, {
  useCallback, useMemo, useRef, useState,
} from 'react';
import { useNavigation } from '@react-navigation/native';
import Carousel, { ICarouselInstance } from 'react-native-reanimated-carousel';
import { Pressable, View } from 'react-native';
import { Box } from '@usereservaapp/reserva-ui';
import { useSharedValue } from 'react-native-reanimated';
import testProps from '../../../../utils/testProps';
import configDeviceSizes from '../../../../utils/configDeviceSizes';
import type { HomeCarouselItemOutput, HomeCarouselOutput } from '../../../../base/graphql/generated';
import ImageComponent from '../../../../components/ImageComponent/ImageComponent';
import { COLORS } from '../../../../base/styles/colors';
import { styles } from './HomeMainCarousel.styles';
import CarrouselScrollIndicator from '../../../../modules/Home/component/CarouselScrollIndicator';
import CarouselPaginationItem from '../../../../components/CarouselPaginationItem';
import { useRemoteConfig } from '../../../../hooks/useRemoteConfig';

interface IHomeMainCarousel {
  data: HomeCarouselOutput;
}

function HomeMainCarousel({ data }: IHomeMainCarousel) {
  const navigation = useNavigation();
  const progressValue = useSharedValue<number>(0);
  const { getBoolean } = useRemoteConfig();

  const [currIndex, setCurrIndex] = useState(0);
  const $carousel = useRef<ICarouselInstance>();

  const slideDelay = useMemo(
    () => (data.showtime || 10) * 1000,
    [data.showtime],
  );

  const showNewHome = getBoolean('show_new_home');

  const carouselHeight = useMemo(() => {
    const [item] = data.items;

    const { width, height } = item!.image;

    return width && height
      ? (height * configDeviceSizes.DEVICE_WIDTH) / width
      : 500;
  }, [data.items]);

  const onPress = useCallback(
    (item: HomeCarouselItemOutput) => {
      const { reservaMini, orderBy } = item;

      return navigation.navigate('ProductCatalog', {
        facetInput: item.facets,
        referenceId: item.reference,
        reservaMini,
        orderBy,
      });
    },
    [navigation],
  );

  return (
    <View style={{ flex: 1 }} {...testProps('default_carrousel_container')}>
      <Carousel
        {...testProps('default_carrousel_content')}
        loop
        width={configDeviceSizes.DEVICE_WIDTH}
        height={carouselHeight}
        ref={(carousel) => {
          if (carousel) $carousel.current = carousel;
        }}
        mode="parallax"
        modeConfig={{
          parallaxScrollingScale: 1,
          parallaxScrollingOffset: 50,
        }}
        enabled={data.items.length > 1}
        onProgressChange={(_, absoluteProgress) => {
          progressValue.value = absoluteProgress;
        }}
        panGestureHandlerProps={{ activeOffsetX: [-10, 10] }}
        data={data.items}
        style={{ backgroundColor: 'rgba(0, 0, 0, 0)', position: 'relative' }}
        onSnapToItem={setCurrIndex}
        renderItem={({ item }) => (
          <Box alignItems="flex-start" bg="white">
            <Box mb="quarck" width={1}>
              <Pressable
                {...testProps('carrousel_button')}
                onPress={() => onPress(item)}
                delayLongPress={100}
              >
                <ImageComponent
                  style={[
                    {
                      height: carouselHeight,
                    },
                    showNewHome
                      ? {
                        borderBottomLeftRadius: 8,
                        borderBottomRightRadius: 8,
                      }
                      : null,
                  ]}
                  resizeMode="cover"
                  source={{ uri: item.image.url }}
                />
              </Pressable>
            </Box>
          </Box>
        )}
      />

      {showNewHome ? (
        <View
          style={[styles.bulletsWrapper, { width: data.items.length * 24.5 }]}
        >
          {data.items.map((item, i) => (
            <CarouselPaginationItem
              backgroundColor={COLORS.WHITE}
              animValue={progressValue}
              index={i}
              actualPosition={currIndex}
              key={`home-main-carousel-${item.image.url}`}
              length={data.items.length}
            />
          ))}
        </View>
      ) : (
        <CarrouselScrollIndicator
          carouselLength={data.items.length}
          actualPosition={currIndex}
          slideDelay={slideDelay}
          onFinishAnimation={() => $carousel.current?.next()}
        />
      )}
    </View>
  );
}

export default HomeMainCarousel;
