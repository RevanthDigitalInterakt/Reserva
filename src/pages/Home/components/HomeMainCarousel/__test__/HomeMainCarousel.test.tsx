import React from 'react';
import {
  render,
  screen,
} from '@testing-library/react-native';
import { theme } from '@usereservaapp/reserva-ui';
import { ThemeProvider } from 'styled-components/native';
import HomeMainCarousel from '../HomeMainCarousel';
import type { HomeCarouselOutput } from '../../../../../base/graphql/generated';

// MOCKS
const mockNavigate = jest.fn();

const mockedCardCarousel = {
  id: 'MAIN',
  type: 'MAIN',
  showtime: 10,
  items: [{
    mkt: false,
    linkMktIn: null,
    reservaMini: false,
    reference: 'collection:2894',
    orderBy: 'RELEVANCIA',
    facets: [{ key: 'productClusterIds', value: '2894', __typename: 'ProductFacetOutput' }],
    image: {
      url: 'https://images.ctfassets.net/6jsfqc13oxv4/RYtXBnHIjoNdgmEvoO9Km/d06af9b5124eca415b7dfbf56fad7163/banner-home-app-liqui_para_todos.gif', title: 'Liqui para Todos Agosto 2023', height: 1400, width: 1080, __typename: 'HomeCarouselItemImageOutput',
    },
  }, {
    mkt: false,
    linkMktIn: null,
    reservaMini: false,
    reference: 'collection:2648',
    orderBy: 'RELEVANCIA',
    facets: [{ key: 'productClusterIds', value: '2648', __typename: 'ProductFacetOutput' }],
    image: {
      url: 'https://images.ctfassets.net/6jsfqc13oxv4/4TFpt2F0YWUdrb2lZ2SYak/09ad88e0d507be9079da04be0733bd05/banner-home-app-od_ate129.jpg', title: 'Camisetas atÃ© R$129', height: 1400, width: 1080, __typename: 'HomeCarouselItemImageOutput',
    },
  }, {
    mkt: false,
    linkMktIn: null,
    reservaMini: false,
    reference: 'collection:2772',
    orderBy: 'RELEVANCIA',
    facets: [{ key: 'productClusterIds', value: '2772', __typename: 'ProductFacetOutput' }],
    image: {
      url: 'https://images.ctfassets.net/6jsfqc13oxv4/2PZVKTj8T4Vkf0rGIw3NsK/2a8f32e7fe4e14ceaec20157001d328f/banner-home-app-ofertas_ate60.gif', title: 'Home 2772 Ofertas Reserva', height: 1400, width: 1080, __typename: 'HomeCarouselItemImageOutput',
    },
  }],
};

jest.mock('react-native-share', () => ({
  open: jest.fn(() => Promise.resolve()),
}));

jest.mock('@react-navigation/native', () => ({
  useNavigation: () => ({ navigate: mockNavigate }),
}));

const TestingComponent = (
  <ThemeProvider theme={theme}>
    <HomeMainCarousel data={mockedCardCarousel as HomeCarouselOutput} />
  </ThemeProvider>
);

describe('HomeMainCarousel', () => {
  beforeEach(async () => {
    jest.useFakeTimers({ legacyFakeTimers: true });
    render(TestingComponent);
  });

  it('should render properly', () => {
    const card = screen.getByTestId('com.usereserva:id/default_carrousel_content');
    expect(card).toBeOnTheScreen();
  });

  it('should match to snapshot', () => {
    expect(screen.toJSON()).toMatchSnapshot();
  });
});
