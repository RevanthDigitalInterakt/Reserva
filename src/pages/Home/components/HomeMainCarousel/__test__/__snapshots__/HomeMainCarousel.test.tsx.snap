// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeMainCarousel should match to snapshot 1`] = `
<View
  accessibilityLabel="com.usereserva:id/default_carrousel_container"
  style={
    Object {
      "flex": 1,
    }
  }
  testID="com.usereserva:id/default_carrousel_container"
>
  <View
    collapsable={false}
    onGestureHandlerEvent={[Function]}
    onGestureHandlerStateChange={[Function]}
    onTouchEnd={[Function]}
    onTouchStart={[Function]}
    style={
      Array [
        Object {
          "overflow": "hidden",
        },
        Object {
          "height": 972,
          "width": 750,
        },
        Object {
          "backgroundColor": "rgba(0, 0, 0, 0)",
          "position": "relative",
        },
        Object {
          "flexDirection": "row",
        },
      ]
    }
    testID="com.usereserva:id/default_carrousel_content"
  >
    <View
      style={
        Array [
          Object {
            "height": 972,
            "position": "absolute",
            "width": 750,
          },
          Object {
            "transform": Array [
              Object {
                "translateX": undefined,
              },
              Object {
                "scale": undefined,
              },
            ],
            "zIndex": undefined,
          },
        ]
      }
    />
    <View
      style={
        Array [
          Object {
            "height": 972,
            "position": "absolute",
            "width": 750,
          },
          Object {
            "transform": Array [
              Object {
                "translateX": undefined,
              },
              Object {
                "scale": undefined,
              },
            ],
            "zIndex": undefined,
          },
        ]
      }
    />
    <View
      style={
        Array [
          Object {
            "height": 972,
            "position": "absolute",
            "width": 750,
          },
          Object {
            "transform": Array [
              Object {
                "translateX": undefined,
              },
              Object {
                "scale": undefined,
              },
            ],
            "zIndex": undefined,
          },
        ]
      }
    />
  </View>
  <View
    accessibilityLabel="com.usereserva:id/carrousel_scroll_indicator_container"
    bottom={10}
    flexDirection="row"
    position="absolute"
    style={
      Array [
        Object {
          "bottom": 10,
          "flexDirection": "row",
          "position": "absolute",
          "zIndex": 3,
        },
      ]
    }
    testID="com.usereserva:id/carrousel_scroll_indicator_container"
    zIndex={3}
  >
    <View
      backgroundColor="neutroFrio2"
      flex={1}
      height={2}
      marginLeft="quarck"
      marginRight="quarck"
      onLayout={[Function]}
      style={
        Array [
          Object {
            "backgroundColor": "#8A8C8E",
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
            "height": 2,
            "marginLeft": 4,
            "marginRight": 4,
          },
          Object {
            "overflow": "hidden",
          },
        ]
      }
    >
      <View
        accessibilityLabel="com.usereserva:id/carrousel_scroll_indicator_animated_view"
        style={
          Object {
            "backgroundColor": "#D8D9DA",
            "height": "100%",
            "left": 0,
            "position": "absolute",
            "top": 0,
            "transform": Array [
              Object {
                "translateX": -10000,
              },
            ],
            "width": "100%",
          }
        }
        testID="com.usereserva:id/carrousel_scroll_indicator_animated_view"
      />
    </View>
    <View
      backgroundColor="neutroFrio2"
      flex={1}
      height={2}
      marginLeft={null}
      marginRight="quarck"
      onLayout={[Function]}
      style={
        Array [
          Object {
            "backgroundColor": "#8A8C8E",
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
            "height": 2,
            "marginRight": 4,
          },
          Object {
            "overflow": "hidden",
          },
        ]
      }
    />
    <View
      backgroundColor="neutroFrio2"
      flex={1}
      height={2}
      marginLeft={null}
      marginRight="quarck"
      onLayout={[Function]}
      style={
        Array [
          Object {
            "backgroundColor": "#8A8C8E",
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
            "height": 2,
            "marginRight": 4,
          },
          Object {
            "overflow": "hidden",
          },
        ]
      }
    />
  </View>
</View>
`;
